### PR TITLE
Increase buffer size to reduce dropped logs

### LINKED ISSUE #1154: Scale testing for Kubearmor Relay
## Request

**Short Description**
We need to perform scale testing for Kubearmor Relay to ensure that it can handle a high volume of traffic without degrading performance or crashing. The goal of this testing is to determine the maximum capacity of the system and identify any bottlenecks that need to be addressed.

**Isuues Found**

- We are dropping events in PushLog() and PushMsg() in kubearmor if all the channels are busy 
- We are dropping events in Relay server if all the channels for client 
- Relay server not consuming events from grpc stream fast enough  
- Buffer size is a bottleneck in relay server causing event drops. 
- Memory utilization of relay for 100 node cluster is alot  

**Tasks**
- [x] Metric when overload relay server
- [x] Increasing the buffer size to reduce dropping of logs 
-  [x] Debug  why the events sent from kubearmor are not equal to events received in relay-server
- [x] Develop a buffer channel to receive events from grpc stream and separate logic to add events to log/alert/msgstruct.
- [x] Determine the number of nodes that the relay server can handle requests from.
- [ ] Make the buffer to receive events dynamic by fetching the number of Kubearmor and multiplying it with the max burst size .
- [ ] Identify a way to manually release the buffer memory for a 100-node cluster with high memory utilization

Untitled-2023-03-17-0128

### PR DESCRIPTION
**Pull request by Shreyas220**
**Purpose of PR?**:

Fixes #1154 

- To address the potential issue of dropped logs when pushing logs into the channels , we increased the buffer size of the `LogStruct.Broadcast`, `AlertStruct.Broadcast`, and `MsgStruct.Broadcast` channels. This will allow the channels to hold more logs and alerts, reducing the likelihood of dropped data
 - Add logging to indicate if a log is dropped or not

### CODE REVIEW

File: KubeArmor/feeder/feeder.go [hunk 1]
```diff
@@ -551,11 +551,18 @@ func (fd *Feeder) PushMessage(level, message string) {
...
+	counter := 0
+	LenMsg := len(MsgStructs)
 	for uid := range MsgStructs {
 		select {
 		case MsgStructs[uid].Broadcast <- &pbMsg:
 		default:
+			counter++
+			if counter == LenMsg {
+				//Default on the last uid in Messagestruct means the msg isnt pushed into Broadcast
+				kg.Print(fmt.Sprint("All channels in PushMessage busy Message dropped, Msg:", pbMsg))
```
Comment (nyrahul) [inline KubeArmor/feeder/feeder.go]:
```suggestion
				kg.Printf("msg channel busy, msg dropped, msg:%+v", pbMsg))
```

File: KubeArmor/feeder/feeder.go [hunk 2]
```diff
@@ -656,11 +663,19 @@ func (fd *Feeder) PushLog(log tp.Log) {
...
+		lenAlert := len(AlertStructs)
 
 		for uid := range AlertStructs {
 			select {
 			case AlertStructs[uid].Broadcast <- &pbAlert:
 			default:
+				counter++
+				if counter == lenAlert {
+					//Default on the last uid in Alterstruct means the Alert isnt pushed into Broadcast
+					kg.Print(fmt.Sprint("All channels in Pushlog busy Alert dropped, Alert:", pbAlert))
```
Comment (nyrahul) [inline KubeArmor/feeder/feeder.go]:
```suggestion
					kg.Printf("log channel busy, alert dropped. alert:%+v", pbAlert))
```

File: KubeArmor/feeder/feeder.go [hunk 3]
```diff
@@ -703,11 +718,17 @@ func (fd *Feeder) PushLog(log tp.Log) {
...
+		counter := 0
+		lenlog := len(LogStructs)
 		for uid := range LogStructs {
 			select {
 			case LogStructs[uid].Broadcast <- &pbLog:
 			default:
+				counter++
+				if counter == lenlog {
+					//Default on the last uid in Logstuct means the log isnt pushed into Broadcase
+					kg.Print(fmt.Sprint("All channels in Pushlog busy Log dropped, Log:", pbLog))
```
Comment (nyrahul) [inline KubeArmor/feeder/feeder.go]:
```suggestion
					kg.Printf("log channel busy, log dropped. log:%+v", pbLog))
```

**Review by nyrahul — CHANGES_REQUESTED** [review]

File: KubeArmor/feeder/feeder.go [hunk 4]
```diff
@@ -116,7 +116,7 @@ func (ls *LogService) removeMsgStruct(uid string) {
 // WatchMessages Function
 func (ls *LogService) WatchMessages(req *pb.RequestMessage, svr pb.LogService_WatchMessagesServer) error {
 	uid := uuid.Must(uuid.NewRandom()).String()
-	conn := make(chan *pb.Message)
+	conn := make(chan *pb.Message, 100)
```
Comment (nyrahul) [inline KubeArmor/feeder/feeder.go]: Pleae keep a constant.

```go
const int QueueSize = 100
```

**Review by nyrahul — CHANGES_REQUESTED** [review]

File: KubeArmor/feeder/feeder.go [hunk 5]
```diff
@@ -551,11 +553,18 @@ func (fd *Feeder) PushMessage(level, message string) {
 
 	MsgLock.Lock()
 	defer MsgLock.Unlock()
-
+	counter := 0
+	LenMsg := len(MsgStructs)
```
Comment (Ankurk99) [inline KubeArmor/feeder/feeder.go]: Since it's not being exported, can we instead use lower-camel case `lenMsg`?

**Review by Ankurk99 — CHANGES_REQUESTED** [review]

**Comment by daemon1024** [comment]
Please rebase and squash the PR. Thanks.

**Review by daemon1024 — APPROVED** [review]

**Review by nyrahul — APPROVED** [review]
LGTM

**Pull request merged**